# 2. Data Preparation

In this notebook we aim to do two things with the raw data `data/all_car_adverts.csv`.

1. Clean the data by fixing the issues we identified in our initial exploration.
2. Transform some of the columns such that they can be used in a machine learning model, i.e. Feature Engineering.


In [1]:
import pandas as pd
import numpy as np
import seaborn as sns

In [2]:
df = pd.read_csv("../data/all_car_adverts.csv")
df.head()

,Unnamed: 0,make,model,variant,car_price,car_badges,car_title,car_sub_title,car_attention_grabber,car_specs,...,num_owner,ulez,full_service,part_service,part_warranty,full_dealership,first_year_road_tax,brand_new,finance_available,discounted
0,0,AC,Cobra,NaN,89995.0,NaN,AC Cobra,4.9 MK IV CRS 2dr,GENUINE AC COBRA CRS 522 BHP,"2001 (X reg), Convertible, 14,400 miles, 4.9L,...",...,5.0,0,0,0,0,0,0,0,0,0
1,1,AC,Cobra,NaN,92500.0,'',AC Cobra,378 - MkIV,PHYSICAL CAR!,"2019 (T reg), Convertible, 650 miles, Manual, ...",...,NaN,0,0,0,0,0,0,0,0,0
2,2,AC,Cobra,NaN,109995.0,'',AC Cobra,MK1V 212 SC. 3.5 V8 350 BHP LOTUS TWIN TURBO. ...,FULL CARBON FIBRE BODY.,"2000 (X reg), Convertible, 21,600 miles, 3.5L,...",...,3.0,0,0,0,0,0,0,0,0,0
3,3,AC,Cobra,NaN,124950.0,'',AC Cobra,302 MKIV 2dr,ABSOLUTELY STUNNING,"1989 (F reg), Convertible, 2,750 miles, Manual...",...,NaN,0,0,0,0,0,0,0,0,0
4,4,AC,Cobra,NaN,124950.0,'',AC Cobra,302 MKIV With Factory Lightweight Engine 5.0 2dr,'STAGE 3' SVO ENGINE,"1989 (E reg), Convertible, 15,142 miles, 5.0L,...",...,NaN,0,0,0,0,0,0,0,0,0


## 2.1. Column Renaming

In this section we clean the dataset and fix as many of the issues that we identified during initial analysis as possible.

#### Redundant Index Column
Firstly, we identified `Unnamed: 0` as an index column. This is a redundant column since we have access to indices through pandas' API anyway.

In [3]:
df = df.drop(columns=["Unnamed: 0"])
df.head()

,make,model,variant,car_price,car_badges,car_title,car_sub_title,car_attention_grabber,car_specs,car_seller,...,num_owner,ulez,full_service,part_service,part_warranty,full_dealership,first_year_road_tax,brand_new,finance_available,discounted
0,AC,Cobra,NaN,89995.0,NaN,AC Cobra,4.9 MK IV CRS 2dr,GENUINE AC COBRA CRS 522 BHP,"2001 (X reg), Convertible, 14,400 miles, 4.9L,...",Private seller,...,5.0,0,0,0,0,0,0,0,0,0
1,AC,Cobra,NaN,92500.0,'',AC Cobra,378 - MkIV,PHYSICAL CAR!,"2019 (T reg), Convertible, 650 miles, Manual, ...",Boss Motor Company Limited,...,NaN,0,0,0,0,0,0,0,0,0
2,AC,Cobra,NaN,109995.0,'',AC Cobra,MK1V 212 SC. 3.5 V8 350 BHP LOTUS TWIN TURBO. ...,FULL CARBON FIBRE BODY.,"2000 (X reg), Convertible, 21,600 miles, 3.5L,...",PREMIER GT LIMITED,...,3.0,0,0,0,0,0,0,0,0,0
3,AC,Cobra,NaN,124950.0,'',AC Cobra,302 MKIV 2dr,ABSOLUTELY STUNNING,"1989 (F reg), Convertible, 2,750 miles, Manual...",TOP 555 LIMITED,...,NaN,0,0,0,0,0,0,0,0,0
4,AC,Cobra,NaN,124950.0,'',AC Cobra,302 MKIV With Factory Lightweight Engine 5.0 2dr,'STAGE 3' SVO ENGINE,"1989 (E reg), Convertible, 15,142 miles, 5.0L,...",TOP 555 LIMITED,...,NaN,0,0,0,0,0,0,0,0,0


#### Column Naming

Due to personal preference, we also change the naming convention of the columns from `snake_case` to `CamelCase`.

In [4]:
renaming_map = {
    x: x.replace("_", " ").title().replace(" ", "")
    for x in df.columns.values
}
df = df.rename(columns=renaming_map)

df.head()

,Make,Model,Variant,CarPrice,CarBadges,CarTitle,CarSubTitle,CarAttentionGrabber,CarSpecs,CarSeller,...,NumOwner,Ulez,FullService,PartService,PartWarranty,FullDealership,FirstYearRoadTax,BrandNew,FinanceAvailable,Discounted
0,AC,Cobra,NaN,89995.0,NaN,AC Cobra,4.9 MK IV CRS 2dr,GENUINE AC COBRA CRS 522 BHP,"2001 (X reg), Convertible, 14,400 miles, 4.9L,...",Private seller,...,5.0,0,0,0,0,0,0,0,0,0
1,AC,Cobra,NaN,92500.0,'',AC Cobra,378 - MkIV,PHYSICAL CAR!,"2019 (T reg), Convertible, 650 miles, Manual, ...",Boss Motor Company Limited,...,NaN,0,0,0,0,0,0,0,0,0
2,AC,Cobra,NaN,109995.0,'',AC Cobra,MK1V 212 SC. 3.5 V8 350 BHP LOTUS TWIN TURBO. ...,FULL CARBON FIBRE BODY.,"2000 (X reg), Convertible, 21,600 miles, 3.5L,...",PREMIER GT LIMITED,...,3.0,0,0,0,0,0,0,0,0,0
3,AC,Cobra,NaN,124950.0,'',AC Cobra,302 MKIV 2dr,ABSOLUTELY STUNNING,"1989 (F reg), Convertible, 2,750 miles, Manual...",TOP 555 LIMITED,...,NaN,0,0,0,0,0,0,0,0,0
4,AC,Cobra,NaN,124950.0,'',AC Cobra,302 MKIV With Factory Lightweight Engine 5.0 2dr,'STAGE 3' SVO ENGINE,"1989 (E reg), Convertible, 15,142 miles, 5.0L,...",TOP 555 LIMITED,...,NaN,0,0,0,0,0,0,0,0,0


We rename `Ulez` to `ULEZ` because it is an acronym and we also rename `FeulType` to `FuelType` to fix the typo.

In [5]:
renaming_map = {
    "Ulez": "ULEZ",
    "FeulType": "FuelType"
}
df = df.rename(columns=renaming_map)

df.head()

,Make,Model,Variant,CarPrice,CarBadges,CarTitle,CarSubTitle,CarAttentionGrabber,CarSpecs,CarSeller,...,NumOwner,ULEZ,FullService,PartService,PartWarranty,FullDealership,FirstYearRoadTax,BrandNew,FinanceAvailable,Discounted
0,AC,Cobra,NaN,89995.0,NaN,AC Cobra,4.9 MK IV CRS 2dr,GENUINE AC COBRA CRS 522 BHP,"2001 (X reg), Convertible, 14,400 miles, 4.9L,...",Private seller,...,5.0,0,0,0,0,0,0,0,0,0
1,AC,Cobra,NaN,92500.0,'',AC Cobra,378 - MkIV,PHYSICAL CAR!,"2019 (T reg), Convertible, 650 miles, Manual, ...",Boss Motor Company Limited,...,NaN,0,0,0,0,0,0,0,0,0
2,AC,Cobra,NaN,109995.0,'',AC Cobra,MK1V 212 SC. 3.5 V8 350 BHP LOTUS TWIN TURBO. ...,FULL CARBON FIBRE BODY.,"2000 (X reg), Convertible, 21,600 miles, 3.5L,...",PREMIER GT LIMITED,...,3.0,0,0,0,0,0,0,0,0,0
3,AC,Cobra,NaN,124950.0,'',AC Cobra,302 MKIV 2dr,ABSOLUTELY STUNNING,"1989 (F reg), Convertible, 2,750 miles, Manual...",TOP 555 LIMITED,...,NaN,0,0,0,0,0,0,0,0,0
4,AC,Cobra,NaN,124950.0,'',AC Cobra,302 MKIV With Factory Lightweight Engine 5.0 2dr,'STAGE 3' SVO ENGINE,"1989 (E reg), Convertible, 15,142 miles, 5.0L,...",TOP 555 LIMITED,...,NaN,0,0,0,0,0,0,0,0,0


## 2.2. Data Types

We notice that some of the data types seem incorrect. For example, `NumOwner` is clearly not an integer column - but cars can only have an integer amount of owners.

In [6]:
df.dtypes

Make                       str
Model                      str
Variant                    str
CarPrice               float64
CarBadges                  str
CarTitle                   str
CarSubTitle                str
CarAttentionGrabber        str
CarSpecs                   str
CarSeller                  str
CarSellerRating        float64
CarSellerLocation          str
Year                       str
Reg                        str
BodyType                   str
Miles                  float64
EngineVol              float64
EngineSize             float64
EngineSizeUnit             str
Transmission               str
FuelType                   str
NumOwner               float64
ULEZ                     int64
FullService              int64
PartService              int64
PartWarranty             int64
FullDealership           int64
FirstYearRoadTax         int64
BrandNew                 int64
FinanceAvailable         int64
Discounted               int64
dtype: object

#### Boolean Columns

Some of the `int64` columns only contain 0s or 1s, and thus can be converted to booleans. We check every `int64` column, and convert it to `bool` iff its min and max values are in $\{0,1\}$.

In [7]:
def check_and_change(col_name):
    col = df[col_name]

    if not pd.api.types.is_integer_dtype(col):
        return False
    
    if col.max() > 1 or col.min() < 0:
        return False
    
    df[col_name] = df[col_name].astype(bool)
    return True


for col_name in df.columns.values:
    if check_and_change(col_name):
        print(f"{col_name} changed from an integer column to a boolean column")

df.head()

ULEZ changed from an integer column to a boolean column
FullService changed from an integer column to a boolean column
PartService changed from an integer column to a boolean column
PartWarranty changed from an integer column to a boolean column
FullDealership changed from an integer column to a boolean column
FirstYearRoadTax changed from an integer column to a boolean column
BrandNew changed from an integer column to a boolean column
FinanceAvailable changed from an integer column to a boolean column
Discounted changed from an integer column to a boolean column


,Make,Model,Variant,CarPrice,CarBadges,CarTitle,CarSubTitle,CarAttentionGrabber,CarSpecs,CarSeller,...,NumOwner,ULEZ,FullService,PartService,PartWarranty,FullDealership,FirstYearRoadTax,BrandNew,FinanceAvailable,Discounted
0,AC,Cobra,NaN,89995.0,NaN,AC Cobra,4.9 MK IV CRS 2dr,GENUINE AC COBRA CRS 522 BHP,"2001 (X reg), Convertible, 14,400 miles, 4.9L,...",Private seller,...,5.0,False,False,False,False,False,False,False,False,False
1,AC,Cobra,NaN,92500.0,'',AC Cobra,378 - MkIV,PHYSICAL CAR!,"2019 (T reg), Convertible, 650 miles, Manual, ...",Boss Motor Company Limited,...,NaN,False,False,False,False,False,False,False,False,False
2,AC,Cobra,NaN,109995.0,'',AC Cobra,MK1V 212 SC. 3.5 V8 350 BHP LOTUS TWIN TURBO. ...,FULL CARBON FIBRE BODY.,"2000 (X reg), Convertible, 21,600 miles, 3.5L,...",PREMIER GT LIMITED,...,3.0,False,False,False,False,False,False,False,False,False
3,AC,Cobra,NaN,124950.0,'',AC Cobra,302 MKIV 2dr,ABSOLUTELY STUNNING,"1989 (F reg), Convertible, 2,750 miles, Manual...",TOP 555 LIMITED,...,NaN,False,False,False,False,False,False,False,False,False
4,AC,Cobra,NaN,124950.0,'',AC Cobra,302 MKIV With Factory Lightweight Engine 5.0 2dr,'STAGE 3' SVO ENGINE,"1989 (E reg), Convertible, 15,142 miles, 5.0L,...",TOP 555 LIMITED,...,NaN,False,False,False,False,False,False,False,False,False


9 columns were changed by this process:

1. `ULEZ`
2. `FullService`
3. `PartService`
4. `PartWarranty`
5. `FullDealership`
6. `FirstYearRoadTax`
7. `BrandNew`
8. `FinanceAvailable`
9. `Discounted`

#### Converting `NumOwner` to Integer

We convert the column `NumOwner` to integer.

In [8]:
df["NumOwner"] = df["NumOwner"].astype(pd.Int64Dtype())
df[["Make", "Model", "Variant", "NumOwner"]].head(20)

,Make,Model,Variant,NumOwner
0,AC,Cobra,NaN,5
1,AC,Cobra,NaN,<NA>
2,AC,Cobra,NaN,3
3,AC,Cobra,NaN,<NA>
4,AC,Cobra,NaN,<NA>
5,AC,Cobra,NaN,<NA>
6,AC,NaN,MK IV,<NA>
7,AC,NaN,MK IV CRS,5
8,Abarth,124 Spider,GT,2
9,Abarth,124 Spider,GT,2


#### Converting `Year` to Integer

Currently `Year` is a string column. From initial analysis, we know that this is because some entries into the column are erroneous and are misentries belonging to different columns such as `Miles` or `BodyType`. For now, we simply convert to integer, like `NumOwners` and coerce any non-numeric value into NaN. We will concern ourselves with the erroneous numeric entries later.

In [9]:
df["Year"] = pd.to_numeric(df["Year"], errors="coerce").astype(pd.Int64Dtype())
df[["Make", "Model", "Variant", "Year"]].tail(20)

,Make,Model,Variant,Year
818436,Volvo,XC90,h T8,<NA>
818437,Volvo,XC90,h T8,2022
818438,Volvo,XC90,h T8,<NA>
818439,Volvo,XC90,h T8,2022
818440,Volvo,XC90,h T8,<NA>
818441,Volvo,XC90,h T8,<NA>
818442,Westfield,Roadster,NaN,2020
818443,Westfield,SE,NaN,1999
818444,Westfield,SE,NaN,1991
818445,Westfield,SEight,NaN,1998


## 2.3. Data Recovery

In the initial analysis, we noted that some of the data in the columns are missing. We attempt to recover as much as this as possible automatically.

#### `Model` Recovery

Some rows are missing their `Model` value. Although initial analysis showed that in most cases `CarTitle` is simply a string concatenation of `Make` and `Model`, we cannot reliably recover the data using the column. Hence we take a different approach. There are a lot of duplicate (`Make`, `Model`) pairs in the dataset, many of these also have the same `CarTitle` value. Hence we do the following.

- Construct a map `CarTitle` $\rightarrow$ `Model` using the data in the database, by grouping by `CarTitle` and taking the modal `Model`.
- Use this to fill in missing `Model` entries where possible.
- Manually inspect what remains.

In [10]:
title_to_model = df.groupby("CarTitle")["Model"].apply(
    lambda s: s.mode().iloc[0] if not s.mode().empty else pd.NA
)
title_to_model

CarTitle
AC Cobra                        Cobra
AK Cobra                        Cobra
Abarth 124                 124 Spider
Abarth 124 SPIDER          124 Spider
Abarth 124 Spider          124 Spider
                            ...      
Westfield Seight               SEight
Westfield Sport                 Sport
Wolseley 1500                    1500
Yamaha YXZ 1000R            Yxz 1000r
Zimmer Golden Spirit    Golden Spirit
Name: Model, Length: 3946, dtype: str

In [11]:
mask = df["Model"].isna()
pre_map = len(df[mask])

df.loc[
    mask,
    "Model"
] = df.loc[mask, "CarTitle"].map(title_to_model)

mask = df["Model"].isna()
post_map = len(df[mask])

print(f"#NA models pre-map: {pre_map}")
print(f"#NA models post-map: {post_map}")

df[mask][["Make", "Model", "Variant", "CarTitle", "CarSubTitle"]]

#NA models pre-map: 303
#NA models post-map: 2


,Make,Model,Variant,CarTitle,CarSubTitle
5172,Alpine,NaN,Legende GT,Alpine A110 Legende GT,1.8 Turbo Legende GT DCT Euro 6 2dr
654124,Sunbeam,NaN,Unlisted,Sunbeam Alpine,NaN


Automated recovery was successful in all but 2 cases. However we can reasonably recover these models manually by inspecting the columns by eye.

In [12]:
df.loc[5172, "Model"] = "A110"
df.loc[654124, "Model"] = "Alpine"

df[mask][["Make", "Model", "Variant", "CarTitle", "CarSubTitle"]]

,Make,Model,Variant,CarTitle,CarSubTitle
5172,Alpine,A110,Legende GT,Alpine A110 Legende GT,1.8 Turbo Legende GT DCT Euro 6 2dr
654124,Sunbeam,Alpine,Unlisted,Sunbeam Alpine,NaN


In [13]:
print(f"Remaining N/A Model Entries: {df["Model"].isna().sum()}")

Remaining N/A Model Entries: 0


We have successfully recovered `Model` in all cases using a mixture of both automatic and manual techinques.

#### `Variant` (Non-) Recovery

In theory, `Variant` can also be recovered using the `CarTitle` and `CarSubTitle` columns. However, we choose not to here for three reasons:

1. Where the `Variant` value would appear in these columns is not consistent across records, thus an automated method would take a long time to develop since we would have to ensure we are extracting the correct substring.
2. `CarTitle` and `CarSubTitle` pairings may not have many duplicates in the table, meaning a method similar to the one we used to recover `Model` would be inefficient.
3. Initial analysis identified that many entries in the `Variant` column are erroneous (another reason why a `Model`-like method would be unlikely to work) and so it seems that this column is generally unreliable.

For these reasons, we will _not_ recover missing `Variant` entries nor correct the erroneous ones. This therefore means that, unfortunately, we are unable to use this as a feature in modelling.

#### `Year` Recovery

We recover missing `Year` data. However, first we check that all years are reasonable.

In [14]:
summary_cols = ["Make", "Model", "Variant", "Year", "Reg", "CarSpecs"]
df.nlargest(10, "Year")[summary_cols]

,Make,Model,Variant,Year,Reg,CarSpecs
5,AC,Cobra,NaN,2022,22 reg,"2022 (22 reg), Convertible, 500 miles, Manual,..."
437,Abarth,595,F595,2022,NaN,"2022, Hatchback, 1.4L, 167PS, Manual, Petrol, ..."
438,Abarth,595,F595,2022,NaN,"2022, Hatchback, 1.4L, 167PS, Manual, Petrol, ..."
439,Abarth,595,F595,2022,NaN,"2022, Hatchback, 1.4L, 167PS, Manual, Petrol, ..."
440,Abarth,595,F595,2022,NaN,"2022, Hatchback, 1.4L, 167PS, Manual, Petrol, ..."
441,Abarth,595,F595,2022,72 reg,"2022 (72 reg), Hatchback, 7 miles, 1.4L, 162BH..."
442,Abarth,595,F595,2022,72 reg,"2022 (72 reg), Hatchback, 3 miles, 1.4L, 162BH..."
443,Abarth,595,F595,2022,NaN,"2022, Hatchback, 1.4L, 167PS, Manual, Petrol, ..."
445,Abarth,595,F595,2022,72 reg,"2022 (72 reg), Hatchback, 10 miles, 1.4L, 167P..."
449,Abarth,595,F595,2022,72 reg,"2022 (72 reg), Hatchback, 10 miles, 1.4L, 167P..."


In [15]:
df.nsmallest(10, "Year")[summary_cols + ["CarSubTitle", "CarTitle"]]

,Make,Model,Variant,Year,Reg,CarSpecs,CarSubTitle,CarTitle
237167,Ford,Ranger,TDCi,32,NaN,"32,375 miles, 3.2L, 200PS, Automatic, Diesel, ...",RANGER WILDTRAK DOUB 3.2,Ford RANGER 2019.5MY
237824,Ford,Ranger,Wildtrak,32,NaN,"32,375 miles, 3.2L, 200PS, Automatic, Diesel, ...",RANGER WILDTRAK DOUB 3.2,Ford RANGER 2019.5MY
586693,Pontiac,Firebird,NaN,1079,NaN,"1079, 79,000 miles",NaN,Pontiac Firebird
5283,Alvis,Sixteen,NaN,1934,NaN,"1934, Saloon, 65,000 miles, 2.2L, Manual, Petr...",SF16.95 Light Saloon 4dr,Alvis Sixteen
5284,Alvis,Sixteen,NaN,1934,NaN,"1934, Saloon, 65,000 miles, 2.2L, Manual, Petr...",SF16.95 Light Saloon 4dr,Alvis Sixteen
612050,Rolls-Royce,20/25,Unlisted,1934,A reg,"1934 (A reg), Limousine, 18,200 miles, 3.6L, P...",3.6 1934 classic 20/25 4dr,Rolls-Royce 20/25
612051,Rolls-Royce,20/25,Unlisted,1935,NaN,"1935, Saloon, 77,000 miles, Manual, Petrol",Barker 4dr,Rolls-Royce 20/25
378539,MG,TA,NaN,1936,NaN,"1936, Convertible, 65,165 miles, 1.3L, Manual,...",1250 1.3 2dr,MG TA
378538,MG,TA,NaN,1937,NaN,"1937, Convertible, 13,920 miles, 1.3L, Manual,...",1250 1.3,MG TA
612455,Rover,P2,NaN,1938,NaN,"1938, Saloon, 70,000 miles, Manual, Petrol, 3 ...",NaN,Rover P2


Entries into the `Year` column range from 32 to 2022. Of course 2022 is reasonable, but 32 is not. 1079 is not reasonable either. Both 32 entries are Ford Rangers which have 2019 in their `CarTitle`, we can reasonably assume that this is the year related to the car.

In [16]:
df.loc[
    237167,
    "Year"
] = 2019

df.loc[
    237824,
    "Year"
] = 2019

df.loc[[237824, 237167], summary_cols]

,Make,Model,Variant,Year,Reg,CarSpecs
237824,Ford,Ranger,Wildtrak,2019,NaN,"32,375 miles, 3.2L, 200PS, Automatic, Diesel, ..."
237167,Ford,Ranger,TDCi,2019,NaN,"32,375 miles, 3.2L, 200PS, Automatic, Diesel, ..."


We see if we can recover the Pontiac's `Year`, unfortunately this seems much more unlikely since it isn't mentioned in any of the listing columns. Pontiac Firebirds were produced from 1967 to 2002 so there is a reasonable chance that 1079 is a typo and was supposed to be 1979. Nonetheless, we check:

- The `Year` column of other Pontiac Firebirds in the dataset
- If 1079 appears in any of the other columns for this particular Firebird

In [17]:
df[
    (df["Make"] == "Pontiac")
    & (df["Model"] == "Firebird")
].sort_values("Year")[summary_cols + ["CarPrice"]]

,Make,Model,Variant,Year,Reg,CarSpecs,CarPrice
586693,Pontiac,Firebird,NaN,1079,NaN,"1079, 79,000 miles",16500.0
586695,Pontiac,Firebird,NaN,1981,W reg,"1981 (W reg), Coupe, 44,333 miles, 4.9L, Autom...",34995.0
586694,Pontiac,Firebird,NaN,1990,G reg,"1990 (G reg), 77,000 miles",24995.0


In [18]:
df.select_dtypes(include="number").iloc[586693]

CarPrice           16500.0
CarSellerRating        4.0
Year                1079.0
Miles              79000.0
EngineVol             <NA>
EngineSize            <NA>
NumOwner              <NA>
Name: 586693, dtype: Float64

While it's most likely that 1979 is the intended year for the Pontiac Firebird, we cannot recover it with certainty, and thus we will replace it with NA.

In [19]:
df.loc[
    586693,
    "Year"
] = pd.NA

df.loc[[586693], summary_cols]

,Make,Model,Variant,Year,Reg,CarSpecs
586693,Pontiac,Firebird,NaN,<NA>,NaN,"1079, 79,000 miles"


We now inspect the rest of the entries with N/A `Year`. We know from initial analysis that `CarSpecs` will **not** contain the year. This is because we found that `CarSpecs` follows a fixed pattern which is never shuffled, so if a year is N/A or incorrect, it means that the correct year was simply omitted from `CarSpecs`. However, as we found with the Ford Rangers, these years may be present in the `CarTitle` or `CarSubTitle` columns.

In [20]:
recovery_cols = ["Make", "Model", "Variant", "Year", "CarTitle", "CarSubTitle"]

print(f"{len(df[df["Year"].isna()])} rows with Year = NA")

df.loc[
    df["Year"].isna(),
    recovery_cols
].sample(100)

48081 rows with Year = NA


,Make,Model,Variant,Year,CarTitle,CarSubTitle
485920,Mercedes-Benz,Eqc,AMG Line,<NA>,Mercedes-Benz EQC,EQC 400 80kWh AMG Line (Premium) Auto 4MATIC 5dr
691968,Toyota,Yaris Cross,VVT-h,<NA>,Toyota Yaris Cross,1.5 VVT-h Excel E-CVT (s/s) 5dr
373329,MG,Mg Hs,T-GDI,<NA>,MG MG Hs,1.5 T-GDI Excite (s/s) 5dr
171786,Fiat,500E,Icon,<NA>,Fiat 500e,42kWh Icon Hatchback 3dr Electric Auto (118 ps)
594495,Renault,Arkana,E-TECH,<NA>,Renault Arkana,R.S. Line E-TECH Hybrid 145 Auto 1.6 5dr
...,...,...,...,...,...,...
376069,MG,Mg Zs,Exclusive,<NA>,MG ZS,Zs 1.0T GDi Exclusive 5dr DCT Hatchback
427590,Mercedes-Benz,A Class,A200,<NA>,Mercedes-Benz A Class,1.3 A200 AMG Line (Executive) 7G-DCT Euro 6 (s...
184286,Ford,EcoSport,T,<NA>,Ford Ecosport,1.0T EcoBoost Titanium Euro 6 (s/s) 5dr
556951,Nissan,Qashqai,Tekna,<NA>,Nissan Qashqai,Qashqai 1.3 Dig-t MHB 140 Tekna 5dr


There are ~48k rows which do not have a `Year` value. This is far many to check by hand, so we will leave them as NA. This only reflects a very small subset of all dataset records, so if Year seems to be quintessential in predicting price, removing these will not have a grave impact on the project. We keep them for now though.

#### `CarSellerLocation` Standardisation & Recovery

We standardise `CarSellerLocation` to title case, and recover any erroneous cases.

In [21]:
df["CarSellerLocation"] = df["CarSellerLocation"].str.title()

df["CarSellerLocation"].unique()

<ArrowStringArray>
[       'Watford',        'Chesham',     'Pulborough',         'Oakham',
  'Milton Keynes',         'Forres',      'Willesden',       'Alcester',
     'Nottingham',     'Portsmouth',
 ...
     'Newtonmore',    'Nr Amersham',       'Tregaron',      'Stromness',
 'Bridge Of Weir',            '3.1',       'Begbroke',  'Little Hadham',
        'Ibstock',   'Isle Of Skye']
Length: 1939, dtype: str

In [22]:
df[df["CarSellerLocation"] == "3.1"]["CarSeller"].unique()

<ArrowStringArray>
['Parkway Derby']
Length: 1, dtype: str

It seems that some instances of `CarSellerLocation` have value 3.1, which is certainly wrong. All rows with this error all have `CarSeller` "Parkway Derby", which is in Derby.

In [23]:
df.loc[
    df["CarSellerLocation"] == "3.1",
    "CarSellerLocation"
] = "Derby"

df[df["CarSellerLocation"] == "3.1"]["CarSeller"].unique()

<ArrowStringArray>
[]
Length: 0, dtype: str

## 2.4. Calculating New Columns

#### Calculating `EngineBHP`

We introduce a new column into the dataframe which unifies the `EngineSize` and `EngineSizeUnit` columns by converting all engine powers into BHP.

In [24]:
df["EngineSizeUnit"].unique()

<ArrowStringArray>
['bhp', nan, 'ps']
Length: 3, dtype: str

There are exactly 2 units used: BHP and PS. They are related as follows: 

$$
    1 \text{ PS} = 0.98632 \text{ BHP}
$$

When calculating the new column `EngineBHP`, we set values to NaN if either `EngineSize` or `EngineSizeUnit` is NaN.

In [25]:
df["EngineBHP"] = np.nan

bhp = df["EngineSizeUnit"].str.lower().eq("bhp")
ps = df["EngineSizeUnit"].str.lower().eq("ps")

df.loc[bhp, "EngineBHP"] = df.loc[bhp, "EngineSize"]
df.loc[ps, "EngineBHP"] = 0.98632 * df.loc[ps, "EngineSize"]

df[["Make", "Model", "Variant", "EngineSize", "EngineSizeUnit", "EngineBHP"]].head(20)

,Make,Model,Variant,EngineSize,EngineSizeUnit,EngineBHP
0,AC,Cobra,NaN,225.0,bhp,225.0000
1,AC,Cobra,NaN,NaN,NaN,NaN
2,AC,Cobra,NaN,NaN,NaN,NaN
3,AC,Cobra,NaN,NaN,NaN,NaN
4,AC,Cobra,NaN,NaN,NaN,NaN
5,AC,Cobra,NaN,NaN,NaN,NaN
6,AC,Cobra,MK IV,NaN,NaN,NaN
7,AC,Cobra,MK IV CRS,225.0,bhp,225.0000
8,Abarth,124 Spider,GT,167.0,bhp,167.0000
9,Abarth,124 Spider,GT,167.0,bhp,167.0000


## 2.5. Outputting Cleaned Files

We output cleaned data using `parquet` files. We will create 2 files. The first (`data/clean_car_adverts.parquet`) will contain the dataframe as it stands at this current point, i.e. after the cleaning steps above.

In [26]:
df.to_parquet("../data/clean_car_adverts.parquet")

The second file (`data/car_data.parquet`) will contain a stripped down version of the dataframe which only contains data which we wish to use for prediction. We remove any columns which are either redundant or related to the car advert:

- `CarTitle`
- `CarSubtitle`
- `CarSpecs`
- `CarBadges`
- `CarAttentionGrabber`
- `CarSeller`
- `CarSellerRating`
- `FinanceAvailable`
- `Discounted`
- `EngineSize`
- `EngineSizeUnit`

In [27]:
small_df = df.drop(columns=[
    "CarTitle",
    "CarSubTitle",
    "CarSpecs",
    "CarBadges",
    "CarAttentionGrabber",
    "CarSeller",
    "CarSellerRating",
    "FinanceAvailable",
    "Discounted",
    "EngineSize",
    "EngineSizeUnit"
])

small_df.head()

,Make,Model,Variant,CarPrice,CarSellerLocation,Year,Reg,BodyType,Miles,EngineVol,...,FuelType,NumOwner,ULEZ,FullService,PartService,PartWarranty,FullDealership,FirstYearRoadTax,BrandNew,EngineBHP
0,AC,Cobra,NaN,89995.0,Watford,2001,X reg,convertible,14400.0,4.9,...,petrol,5,False,False,False,False,False,False,False,225.0
1,AC,Cobra,NaN,92500.0,Chesham,2019,T reg,convertible,650.0,NaN,...,petrol,<NA>,False,False,False,False,False,False,False,NaN
2,AC,Cobra,NaN,109995.0,Pulborough,2000,X reg,convertible,21600.0,3.5,...,petrol,3,False,False,False,False,False,False,False,NaN
3,AC,Cobra,NaN,124950.0,Oakham,1989,F reg,convertible,2750.0,NaN,...,petrol,<NA>,False,False,False,False,False,False,False,NaN
4,AC,Cobra,NaN,124950.0,Oakham,1989,E reg,convertible,15142.0,5.0,...,petrol,<NA>,False,False,False,False,False,False,False,NaN


In [28]:
small_df.to_parquet("../data/car_data.parquet")

## 2.2. Feature Engineering

In [29]:
cleaned_df = df.copy()

### 2.2.1. Problem Definition

We wish to use the car specifications to attempt to predict the selling price of the car. In simpler terms, we wish to use the data which is used to create the advert, not the advert itself. This means that we do not wish to use columns such as `CarAttentionGrabber` and `CarSeller`.

One column which sits in a bit of a grey area is `CarSellerLocation`, since while it is technically tied to the listing rather than being an intrinsic feature of the car. Geographical location could play into some local supply-demand dynamics. Therefore, we shall leave this column.

This is the complete list of columns dropped due to not being relevant for our particular problem:

- `CarTitle`
- `CarSubtitle`
- `CarSpecs`
- `CarBadges`
- `CarAttentionGrabber`
- `CarSeller`
- `CarSellerRating`
- `FinanceAvailable`
- `Discounted`

In [30]:
cleaned_df.drop(columns=[
    "CarTitle",
    "CarSubTitle",
    "CarSpecs",
    "CarBadges",
    "CarAttentionGrabber",
    "CarSeller",
    "CarSellerRating",
    "FinanceAvailable",
    "Discounted"
], inplace=True)

cleaned_df.head()

,Make,Model,Variant,CarPrice,CarSellerLocation,Year,Reg,BodyType,Miles,EngineVol,...,FuelType,NumOwner,ULEZ,FullService,PartService,PartWarranty,FullDealership,FirstYearRoadTax,BrandNew,EngineBHP
0,AC,Cobra,NaN,89995.0,Watford,2001,X reg,convertible,14400.0,4.9,...,petrol,5,False,False,False,False,False,False,False,225.0
1,AC,Cobra,NaN,92500.0,Chesham,2019,T reg,convertible,650.0,NaN,...,petrol,<NA>,False,False,False,False,False,False,False,NaN
2,AC,Cobra,NaN,109995.0,Pulborough,2000,X reg,convertible,21600.0,3.5,...,petrol,3,False,False,False,False,False,False,False,NaN
3,AC,Cobra,NaN,124950.0,Oakham,1989,F reg,convertible,2750.0,NaN,...,petrol,<NA>,False,False,False,False,False,False,False,NaN
4,AC,Cobra,NaN,124950.0,Oakham,1989,E reg,convertible,15142.0,5.0,...,petrol,<NA>,False,False,False,False,False,False,False,NaN


In [31]:
cleaned_df.drop(columns=[
    "EngineSize",
    "EngineSizeUnit"
], inplace=True)

cleaned_df.head()

,Make,Model,Variant,CarPrice,CarSellerLocation,Year,Reg,BodyType,Miles,EngineVol,...,FuelType,NumOwner,ULEZ,FullService,PartService,PartWarranty,FullDealership,FirstYearRoadTax,BrandNew,EngineBHP
0,AC,Cobra,NaN,89995.0,Watford,2001,X reg,convertible,14400.0,4.9,...,petrol,5,False,False,False,False,False,False,False,225.0
1,AC,Cobra,NaN,92500.0,Chesham,2019,T reg,convertible,650.0,NaN,...,petrol,<NA>,False,False,False,False,False,False,False,NaN
2,AC,Cobra,NaN,109995.0,Pulborough,2000,X reg,convertible,21600.0,3.5,...,petrol,3,False,False,False,False,False,False,False,NaN
3,AC,Cobra,NaN,124950.0,Oakham,1989,F reg,convertible,2750.0,NaN,...,petrol,<NA>,False,False,False,False,False,False,False,NaN
4,AC,Cobra,NaN,124950.0,Oakham,1989,E reg,convertible,15142.0,5.0,...,petrol,<NA>,False,False,False,False,False,False,False,NaN


### 2.2.3. Summary

In this section we have excluded columns from the dataset which are not relevant for our prediction model, and we have cut down the table size by a single column by converting all engine powers into a single unit. We save the resulting dataset to `data/engineered_car_adverts.parquet`.

In [32]:
cleaned_df.to_parquet("../data/engineered_car_adverts.parquet")

## 2.3. Summary

In this notebook, we have cleaned the dataset, and considered what features we wish to use for modelling. We have created two parquet files from the original dataset.

1. `data/clean_car_adverts.parquet` - the resulting output after cleaning, contains both metrics of engine power as well as data relating to the car adverts themselves.
2. `data/engineered_car_adverts.parquet` - the resulting output after feature engineering with a unified engine power column and after removing the irrelevant columns of the dataset.

Some records still have missing data in some columns. However, at this stage, I think it would be a bad idea to consider either deleting more columns, or deleting records with missing data since we have not yet explored which features are essential for price prediction. This is the goal of the next notebook.